### PR TITLE
docs: Add top-level Go applications page

### PR DESCRIPTION
**Pull request by hkjn**
Also move middleware.md into docs/go/ and add some minimal content to
the page.

Also add /docs/go/build.md with a description of the Go build process.

Also add /docs/go/tools.md doc with a minimal description of `bbbfancontrol`
and `bbbsupervisor`.

Also add link to Go docs from docs/os/build-details.md.

### CODE REVIEW

File: docs/go/tools.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+layout: default
+title: Go tools
+nav_order: 300
+parent: Go Services
+permalink: /go/tools
+---
+## BitBox Base: Go tools
+
+We have some Go tools that are part of this repository:
```
Comment (Stadicus) [inline docs/go/tools.md]: Same here. How about:
> Several custom Go tools are part of this repository:

Also: further detailed documentation of these tools in `/docs/go/tools/`?

File: docs/os/build-details.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ It performs the following steps:
 1. cloning the [`github.com/armbian/build`](https://github.com/armbian/build/) repo into `armbian/armbian-build/`, if it doesn't exist already
 1. copying over the following files from the host system into `armbian/armbian-build/userpatches/overlay/`:
     1. `armbian/base/`: scripts and configs included in the Armbian image
-    1. `build/`: contains all built Go binaries and their associated `.service` files
+    1. `build/`: contains all [built Go binaries](/go/) and their associated `.service` files
```
Comment (Stadicus) [inline docs/os/build-details.md]: I think that should be `/docs/go-services.md`

File: docs/go-services.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+---
+layout: default
+title: Go Services
```
[the remark this answers is not in the document]
Comment (hkjn) [reply docs/go-services.md]: I was using "service" as an informal way to refer to a long-running process that serves some need, such as "business logic" or monitoring or other system features. I'm fine using "applications" instead for all the things written in Go that run persistently, if that's the terminology you prefer.

File: docs/go/build.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
+---
+layout: default
+title: Building Go binaries
+nav_order: 100
+parent: Go Services
+permalink: /go/build
+---
+## BitBox Base: Building Go binaries
+
+We have some Go binaries that become part of [the Armbian image](/os/armbian-build.html) that the BitBox Base runs. This page describes the process used to build those images.
```
[the remark this answers is not in the document]
Comment (hkjn) [reply docs/go/build.md]: Done, thanks.

File: docs/go-services.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Stadicus) [reply docs/go-services.md]: Ok, let's keep it simple and stick with "application"

File: docs/go/tools.md (hunk 1 above)
Comment (hkjn) [reply docs/go/tools.md]: > Same here. How about:
> 
> > Several custom Go tools are part of this repository:
I changed the text to `Several Go tools are part of this repository`, thanks.
> 
> Also: further detailed documentation of these tools in `/docs/go/tools/`?

Not sure what this is referring to? We're discussing on the `docs/go/tools.md` document; what text would you like me to add?

File: docs/os/build-details.md (hunk 2 above)
Comment (hkjn) [reply docs/os/build-details.md]: Ah, I was trying to guess / reverse-engineer what URLs Jekyll would render these pages on (that was the reason for adding `permalink` stanzas too, btw), but if we can refer to the `.md` files and let Jekyll sort out what `<a href="...">` values to put in, that's certainly better!

Fixed, thanks.

File: docs/go-services.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hkjn) [reply docs/go-services.md]: I updated the wording to "Go applications".

**Comment by hkjn** [comment]
Thanks for the review, @Stadicus! PTAL.

> Just to make sure I understand the goal of this pull request:

Please see the updated commit description for d4f62e1f77d670f6a6e0195bf1ef11686ef9a495, and let me know if anything remains unclear about the changes made in this PR.

> 
>     * it adds a new `/docs/go/` directory
> 
>     * according to the `permalink`s you add three additional subdirectories under `/docs/go/`
As described in a comment thread, the `permalink`s were there as a workaround to try to create predictable URLs, but I've dropped them at this point and I am now just referencing the `.md` files in the links, letting Jekyll sort out where the `<a href="...">` should point.

> 
>     * it describes the generig Go build process in `/docs/go/build.md` and adds additional sub-chapters in `/docs/go/build/...`
> 
>     * it therefore replaces current chapter 5 (Middleware) and will add all these Middleware detail chapters under `/docs/go/middleware/...`
> 
>     * it also moves current "Custom Tools" (chapter 7.4) with `/docs/go/tools.md` and creates a subdirectory to add detailed documentation in `/docs/go/tools/...`
> 
> 
> I think that's a good approach, as long as
> 
>     * the sub-subdirectories are really necessary (meaning having 3+ additional `.md` files in them), which I'm not sure regarding the Go build documentation
> 
>     * the overall build process is additionally described early on (as currently proposed as chapter 2 "Creating the Base image")
> 
>     * all relevant detail documentation of the Middleware and custom Go tools will become part of this structure.

I'm not 100% clear if you are suggesting the three bulletpointed changes should be made in this PR, or for future work?

**Comment by hkjn** [comment]
> Please see the updated commit description for [d4f62e1](https://github.com/digitalbitbox/bitbox-base/commit/d4f62e1f77d670f6a6e0195bf1ef11686ef9a495), and let me know if anything remains unclear about the changes made in this PR.

Apologies, I had overlooked some references to `Go Services`, which I've now updated to `Go applications` consistently and force-pushed as 8f7c81daacd67afb4f3af72bad7213c327c95027.

File: docs/go/middleware.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+---
+layout: default
+title: Middleware
+nav_order: 200
+parent: Go Services
+permalink: /go/middleware
+---
+## BitBox Base: Middleware
+
+We have a `middleware` service written in Go, which exposes a single endpoint to the network, acting as a server when the user connects using the [BitBox App](https://github.com/digitalbitbox/bitbox-wallet-app/):
```
[the remark this answers is not in the document]
Comment (hkjn) [reply docs/go/middleware.md]: Oops, I missed this reference to "service". Fixed to say "application" in ca72081.

**Comment by Stadicus** [comment]
> * the sub-subdirectories are really necessary (meaning having 3+ additional `.md` files in them), which I'm not sure regarding the Go build documentation

Ok, I have misunderstood your intent to create an additional level of sub-directories. ✅

File: docs/go/middleware.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (hkjn) [reply docs/go/middleware.md]: > Is the plan to include the detailed Middleware documentation, including all sub-chapters on API specification, Noise encryption, HSM etc. here in /docs/go/middleware/?

Sure, unless there's reasons to put it elsewhere.

**Comment by Stadicus** [comment]
> * the overall build process is additionally described early on (as currently proposed as chapter 2 "Creating the Base image")

Not meant for this pull-request, I just wanted to point out that there is an "entry point" missing for people approaching this project to understand how to get started with building the whole thing. Can be a single page overview "3) Creating the Base image" without all the sub-chapters.

**Comment by Stadicus** [comment]
> * all relevant detail documentation of the Middleware and custom Go tools will become part of this structure.

I just want to make sure that we know where the additional detail chapters for the Middleware are supposed to be later on (as you deleted the `middleware.md` chapter). How would you structure it to include "API spec", "Noise Framework", "HSM integration", etc.... ?

**Comment by Stadicus** [comment]
Where would you now put the `supervisor.md` documentation in that structure?

**Comment by hkjn** [comment]
> Not meant for this pull-request, I just wanted to point out that there is an "entry point" missing for people approaching this project to understand how to get started with building the whole thing. Can be a single page overview "3) Creating the Base image" without all the sub-chapters.

Sounds good.

**Comment by hkjn** [comment]
> > * all relevant detail documentation of the Middleware and custom Go tools will become part of this structure.
> 
> I just want to make sure that we know where the additional detail chapters for the Middleware are supposed to be later on (as you deleted the `middleware.md` chapter). How would you structure it to include "API spec", "Noise Framework", "HSM integration", etc.... ?

I would personally add that as subsections to `middleware.md`, or if that grew too large, to more specific documents like `middleware-api.md` etc.

**Comment by hkjn** [comment]
> Where would you now put the `supervisor.md` documentation in that structure?

The `docs/go/tools.md` doc mentions the existence of the `bbbsupervisor`, and links to the source code for it. I'd suggest either adding supervisor-related content directly to `tools.md` or making it into a subpage under `docs/go/`.

No strong feelings here (or in general) on how to structure the documentation, I'd be happy to go with whatever you prefer. :)

**Comment by Stadicus** [comment]
utACK https://github.com/digitalbitbox/bitbox-base/commit/ca7208176ddffcaba5e8fc0e48bf961d831f4b08

**Pull request merged**